### PR TITLE
Async login/register

### PR DESCRIPTION
**Pull request by SalmonDE**
Due to lag spikes mentioned in #56 I opened this pr to add asynchronous login and register.
It is configurable inside the config, so in case people find this unnecessary they don't have to use it.

While most of the login part is performed on the main thread afterwards, the hashing is done in another thread. That's because the hashing was the main cause of the lags and the login sequence after working out the hash needs to use the PocketMine-MP API.

Login is implemented and is ready for testing.
Register still needs to be worked on.

### CODE REVIEW

**Comment by DaPigGuy** [comment]
Has this been tested?

**Comment by SalmonDE** [comment]
Roughly. Using the correct password let me log into my account. Using a false password told me I lost a try.
Can I request testing from others?

File: src/PiggyAuth/Tasks/AsyncLoginTask.php [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+            return;
+        }
+
+        if($this->getResult()['updatePlayer']){
+            $plugin->sessionmanager->getSession($player)->updatePlayer("auth", "PiggyAuth");
+        }
+
+        if (!$this->getResult()['authenticated']) { // Failed
+            if ($plugin->getConfig()->getNested("key.enabled")) {
+                if ($this->potentialPassword === $plugin->getKey($this->potentialPassword)) { // Please look at this @MCPEPIG
```
Comment (MCPEPIG) [inline src/PiggyAuth/Tasks/AsyncLoginTask.php]: This looks fine except the parameter for getKey() has to be the owner's password (owner's username is provided in config)

File: src/PiggyAuth/Tasks/AsyncLoginTask.php [hunk 2]
```diff
@@ -74,6 +74,7 @@ public function needsRehashPassword()
     public function onCompletion(Server $server)
     {
         $plugin = $server->getPluginManager()->getPlugin('PiggyAuth');
+        $plugin->sessionmanager->getSession($player)->setVerifying(); // This should be fine, even if the plugin is disabled, right?
```
Comment (DaPigGuy) [inline src/PiggyAuth/Tasks/AsyncLoginTask.php]: Um you should check if plugin is enabled before doing that....

Comment (SalmonDE) [reply src/PiggyAuth/Tasks/AsyncLoginTask.php]: Fine, I thought it's going to be fine. And wow, I forgot to put in false.

File: resources/config.yml [hunk 3]
```diff
@@ -1,19 +1,19 @@
...
-#|   |   \ :`--'_    /   /  ' / /   /  ' /     .--, :  |   /\   \    .--. |  | :.;__,'  /  :  |  |,--. 
-#|   : .   /,' ,'|  |   :     ||   :     |   /_ ./| |  :  ' ;.   : ,'_ /| :  . ||  |   |   |  :  '   | 
-#;   | |`-' '  | |  |   | .\  .|   | .\  ., ' , ' : |  |  ;/  \   \|  ' | |  . .:__,'| :   |  |   /' : 
-#|   | ;    |  | :  .   ; ';  |.   ; ';  /___/ \: | '  :  | \  \ ,'|  | ' |  | |  '  : |__ '  :  | | | 
-#:   ' |    '  : |__'   .   . |'   .   . |.  \  ' | |  |  '  '--'  :  | : ;  ; |  |  | '.'||  |  ' | : 
-#:   : :    |  | '.'|`---`-'| | `---`-'| | \  ;   : |  :  :        '  :  `--'   \ ;  :    ;|  :  :_:,' 
-#|   | :    ;  :    ;.'__/\_: | .'__/\_: |  \  \  ; |  | ,'        :  ,      .-./ |  ,   / |  | ,'     
-#`---'.|    |  ,   / |   :    : |   :    :   :  \  \`--''           `--`----'      ---`-'  `--''       
-#  `---`     ---`-'   \   \  /   \   \  /     \  ' ;                                                   
-#                      `--`-'     `--`-'       `--`
```
Comment (SalmonDE) [inline resources/config.yml:16]: Sorry for this, my editor removes whitespaces on it's own.

**Comment by SalmonDE** [comment]
Can I turn functions I need in my async tasks to static ones so I don't need to copy them?

**Comment by DaPigGuy** [comment]
Nu

**Comment by SalmonDE** [comment]
Coding seems to be done, testing now.
However, unregister, preregister and changepassword could still lag the server.

**Comment by SalmonDE** [comment]
Seems good so far

**Comment by SalmonDE** [comment]
No issues found. Green light from me.

**Pull request merged**